tensorflow_probability/examples/jupyter_notebooks/Deep_Exponential_Family_TFP.ipynb

```diff
@@ -0,0 +1,1904 @@
+        "$$\n",
+        "\\textbf{Z}_2 = \\text{Gamma}(\\text{shape}=0.1\\text{, scale}=0.1) \\\\\n",
+        "\\textbf{W}_2 = \\text{Gamma}(\\text{shape}=0.1, \\text{scale}=0.3) \\\\\n",
+        "\\textbf{Z}_1 = \\text{Gamma}(\\text{shape}=k\\text{,  scale}=k/(\\textbf{Z}_2\\textbf{W}_2^T)) \\\\\n",
+        "\\textbf{W}_1 = \\text{Gamma}(\\text{shape}=0.1, \\text{scale}=0.3) \\\\\n",
+        "\\textbf{Z}_0 = \\text{Gamma}(\\text{shape}=k\\text{,  scale}=k/(\\textbf{Z}_1\\textbf{W}_1^T)) \\\\\n",
+        "\\textbf{W}_0 = \\text{Gamma}(\\text{shape}=0.1, \\text{scale}=0.3) \\\\\n",
+        "\\text{ }\\\\\n",
+        "x = \\text{Poisson}(\\textbf{Z}_0\\textbf{W}_0^T) \n",
+        "$$"
+      ]
+    },
+    {
+      "metadata": {
+        "id": "zrXhfjr8RHHz",
+        "colab_type": "code",
+        "colab": {}
+      },
+      "cell_type": "code",
+      "source": [
+        "def deep_exponential_family(data_size, feature_size, units, shape):\n",
+        "    \"\"\"A multi-layered topic model over a documents-by-terms matrix.\"\"\"\n",
+        "    w2 = tfd.Gamma(tf.fill([units[2], units[1]], 0.1), tf.fill([units[2], units[1]], 0.3), name=\"w2\")\n",
+        "    w1 = tfd.Gamma(tf.fill([units[1], units[0]], 0.1), tf.fill([units[1], units[0]], 0.3), name=\"w1\")\n",
+        "    w0 = tfd.Gamma(tf.fill([units[0], feature_size], 0.1), tf.fill([units[0], feature_size], 0.3), name=\"w0\")\n",
+        "\n",
+        "    z2 = tfd.Gamma(tf.fill([data_size, units[2]], 0.1), tf.fill([data_size, units[2]], 0.1), name=\"z2\")\n",
+        "    z1 = tfd.Gamma(shape, shape / tf.matmul(z2.sample(), w2.sample()), name=\"z1\")\n",
+        "    z0 = tfd.Gamma(shape, shape / tf.matmul(z1.sample(), w1.sample()), name=\"z0\")\n",
+        "    x = tfd.Poisson(tf.matmul(z0.sample(), w0.sample()), name=\"x\")\n",
+        "    return x\n"
+      ],
+      "execution_count": 0,
+      "outputs": []
+    },
+    {
+      "metadata": {
+        "id": "FF1TvorGnbpP",
+        "colab_type": "text"
+      },
+      "cell_type": "markdown",
+      "source": [
+        "The first of these is our learnable deterministic distribution over positive reals."
+      ]
+    },
+    {
+      "metadata": {
+        "id": "S3ubbvHGCDrF",
+        "colab_type": "code",
+        "colab": {}
+      },
+      "cell_type": "code",
+      "source": [
+        "def trainable_positive_deterministic(shape, min_loc=1e-3, name=None):\n",
+        "    \"\"\"Learnable Deterministic distribution over positive reals.\"\"\"\n",
+        "    with tf.variable_scope(None, default_name=\"trainable_positive_deterministic\"):\n",
+        "        unconstrained_loc = tf.get_variable(\"unconstrained_loc\", shape)\n",
+        "        loc = tf.maximum(tf.nn.softplus(unconstrained_loc), min_loc)\n",
```

Comment:
maximum is not differentiable to the non-max element (i.e. once a value falls below threshold it can never come back). could we just use `softplus(unconstrained)+eps`?